src/ORM/Query.php

```diff
@@ -0,0 +1,240 @@
+    {
+        parent::__construct($connection, $table);
+
+        $this->_paginator = Paginator::create($this);
+    }
+
+    /**
+     * Create query based on the existing query. This factory copies the internal
+     * state of the given query to a new instance.
+     *
+     * @param BaseQuery $query
+     */
+    public static function fromQuery($query)
+    {
+        $obj = new static($query->getConnection(), $query->getRepository());
+
+        foreach (get_object_vars($query) as $k => $v) {
+            $obj->$k = $v;
+        }
+
+        $obj->_executeOrder($obj->clause('order'));
+        $obj->_executeLimit($obj->clause('limit'));
+
+        return $obj;
+    }
+
+    /**
+     * Set the cursor.
+     *
+     * @param Cursor[]|int[]|string[] $cursor
+     */
+    public function cursor($cursor = [])
+    {
+        $this->_cursor = $cursor;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function order($fields, $overwrite = false)
+    {
+        parent::order($fields, $overwrite);
+        $this->_executeOrder($this->clause('order'));
+
+        return $this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function orderAsc($field, $overwrite = false)
+    {
+        parent::orderAsc($field, $overwrite);
+        $this->_executeOrder($this->clause('order'));
+
+        return $this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function orderDesc($field, $overwrite = false)
+    {
+        parent::orderDesc($field, $overwrite);
+        $this->_executeOrder($this->clause('order'));
+
+        return $this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function limit($num)
+    {
+        parent::limit($num);
+        $this->_executeLimit($this->clause('limit'));
+
+        return $this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function all()
+    {
+        return $this->_paginator->paginate($this->_cursor);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function _performCount()
+    {
+        return $this->_paginator->build($this->_cursor)->_performCount();
+    }
+
+    /**
+     * @param  null|OrderByExpression $order
+     * @return void
+     */
+    protected function _executeOrder($order)
+    {
+        $this->_paginator->clearOrderBy();
+
+        if ($order === null) {
+            return;
+        }
+
+        $generator = $this->getValueBinder();
+        $order->iterateParts(function ($condition, $key) use ($generator) {
+            if (!is_int($key)) {
+                /**
+                 * @var string $key       The column
+                 * @var string $condition The order
+                 */
+                $this->_paginator->orderBy($key, $condition);
+            }
+
+            if ($condition instanceof OrderClauseExpression) {
+                $generator->resetCount();
+
+                if (!preg_match('/ (?<direction>ASC|DESC)$/', $condition->sql($generator), $matches)) {
```

Comment:
I'm not familiar with CakePHP 3, but case-insensitive matching looks better.

Reply:
The internal implementation of the `\Cake\Database\Expression\OrderClauseExpression` already normalizes the direction to uppercase.
https://github.com/cakephp/cakephp/blob/3.8.8/src/Database/Expression/OrderClauseExpression.php#L43